lib/capistrano/configuration/scm_resolver.rb

```diff
@@ -0,0 +1,149 @@
+module Capistrano
+  class Configuration
+    # In earlier versions of Capistrano, users would specify the desired SCM
+    # implementation using `set :scm, :git`, for example. Capistrano would then
+    # load the matching .rb file based on this variable.
+    #
+    # Now we expect users to explicitly `require` and call `new` on the desired
+    # SCM implementation in their Capfile. The `set` technique is deprecated.
+    #
+    # This SCMResolver class takes care of managing the transition from the old
+    # to new system. It maintains the legacy behavior, but prints deprecation
+    # warnings when it is used.
+    #
+    # To maintain backwards compatibility, the resolver will load the Git SCM by
+    # if default it determines that no SCM has been explicitly specified or
+    # loaded. To force no SCM to be used at all, use `set :scm, nil`. This hack
+    # won't be necessary once backwards compatibility is removed in a future
+    # version.
+    #
+    # TODO: Remove this class entirely in Capistrano 4.0.
+    #
+    class SCMResolver
+      DEFAULT_GIT = :"default-git"
+
+      include Capistrano::DSL
+
+      def resolve
+        return if scm_name.nil?
+        set(:scm, :git) if using_default_scm?
+
+        print_deprecation_warnings_if_applicable
+        return if scm_plugin_loaded?
+
+
+        if built_in_scm_name?
+          load_built_in_scm
+        else
+          # Compatibility with existing 3.x third-party SCMs
+          register_legacy_scm_hooks
+          load_legacy_scm_by_name
+        end
+      end
+
+      private
+
+      def using_default_scm?
+        return @using_default_scm if defined? @using_default_scm
+        @using_default_scm = (fetch(:scm) == DEFAULT_GIT)
```

Comment:
What I am trying to accomplish here is testing whether `set :scm` was ever called. I do that by initializing `:scm` to a dummy value (`DEFAULT_GIT`) and then testing it like this:

```ruby
# if true, then :scm wasn't changed by the user
fetch(:scm) == DEFAULT_GIT
```

However, for the rest of the SCM code to work, and to not break backwards compatibility with other code that is using the `:scm` value, I have to change `:scm` to equal `:git` later on in the code flow, here: https://github.com/capistrano/capistrano/pull/1572/files#diff-f1f4ccb7067eef4a3d33b71bf7ecc2f8R29

That is why I record the value of the test in `@using_default_scm`, because `fetch(:scm)` will change later. I agree this is a hack, but I couldn't think of a better approach.

Any ideas?